### ISSUE TITLE
Auto-select "software" as the rendering engine with X11 forwarding 

### ISSUE THREAD

**Issue by arnonerba** [rstudio/rstudio#4640]
### System details

    RStudio Edition : Desktop
    RStudio Version : 1.2.1335
    OS Version      : RHEL 7.6 / Ubuntu 18.04
    R Version       : 3.5.2 / 3.5.3

### Describe the problem in detail

When launched over SSH with X11 forwarding enabled, RStudio opens as an empty white window. Setting `desktop.renderingEngine=software` in the per-user `desktop.ini` config file allows RStudio to launch successfully with "software" selected as the rendering engine.

According to [this article](https://support.rstudio.com/hc/en-us/articles/360017886674-Troubleshooting-RStudio-Rendering-Errors), setting the following environment variable should have the same effect as setting `desktop.renderingEngine=software`:
```
export RSTUDIO_CHROMIUM_ARGUMENTS="--disable-gpu"
```
However, setting that environment variable has no effect, and RStudio still launches as an empty white window. I am looking for a way to disable GPU rendering for all users of RStudio on the system.

**TL;DR:** Is `export RSTUDIO_CHROMIUM_ARGUMENTS="--disable-gpu"` not the same as setting `desktop.renderingEngine=software`?

**Comment by jmcphers** [comment]
It's `--enable-webgl-software-rendering`. 

https://github.com/rstudio/rstudio/blob/8439ce32c2cfe66d72aa410b50dd6abee659651c/src/cpp/desktop/DesktopMain.cpp#L394-L402

Does that do what you want?

**Comment by arnonerba** [comment]
Hm. It still doesn't seem to work. The environment variable is set, but RStudio still doesn't launch correctly.
```shell
[user@server:~]$ echo $RSTUDIO_CHROMIUM_ARGUMENTS
--enable-webgl-software-rendering
```

**Comment by kevinushey** [comment]
Does the following work in a terminal?

```
export RSTUDIO_CHROMIUM_ARGUMENTS="--disable-gpu"
/usr/lib/rstudio/bin/rstudio
```

If not, then there might be something else going on.

Also worth confirming: how are you launching RStudio? If you're launching it with e.g. the OS application launcher, environment variables set in the terminal won't take effect. There's usually an OS-specific way of ensuring applications are launched with certain environment variables set but it differs from platform to platform.

**Comment by arnonerba** [comment]
@kevinushey Unfortunately, that doesn't seem to have any effect either. I'm launching RStudio directly from a terminal over an SSH session with X11 forwarding enabled. RStudio is installed on a few headless multi-user servers but has been happy to run over SSH so far until the new rendering engine in 1.2.

For what it's worth, other apps (Firefox) have had issues with hardware accelerated rendering on these same servers. Rather than troubleshoot hardware accelerated video over SSH, I'd like to just roll back to software rendering system-wide, at least for now.

Thanks for your help with this.

**Comment by arnonerba** [comment]
I did some more testing, and it seems that `--disable-gpu` does actually have an effect. However, it still doesn't allow RStudio to launch successfully.

1. Output with no environment variables or settings (white screen on launch):
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Unrecognized OpenGL version
Unrecognized OpenGL version
[201972:202034:0412/161509.009059:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.009101:ERROR:gpu_info_collector.cc(68)] gl::GLContext::MakeCurrent() failed
[201972:202034:0412/161509.009141:ERROR:gpu_info_collector.cc(186)] Could not create context for info collection.
Unrecognized OpenGL version
Unrecognized OpenGL version
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
QQuickWidget: Failed to make context current
QQuickWidget::resizeEvent() no OpenGL context
QQuickWidget: Failed to make context current
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
Unrecognized OpenGL version
Unrecognized OpenGL version
Unrecognized OpenGL version
Unrecognized OpenGL version
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
QQuickWidget: Attempted to render scene with no context
[201972:202034:0412/161509.522086:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.522104:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.523970:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.523999:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.526388:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.526398:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.534254:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.534279:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.537767:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.537780:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.540181:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.540191:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.546920:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.546930:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.550324:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.550333:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.551818:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.551830:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
[201972:202034:0412/161509.553241:ERROR:gl_context_glx.cc(232)] Couldn't make context current with X drawable.
[201972:202034:0412/161509.553249:ERROR:gles2_command_buffer_stub.cc(312)] ContextResult::kTransientFailure: Failed to make context current.
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
QQuickWidget: Attempted to render scene with no context
QQuickWidget::invalidateRenderControl could not make context current
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
```
2. Output with `export RSTUDIO_CHROMIUM_ARGUMENTS="--disable-gpu"` set (also white screen on launch):
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Unrecognized OpenGL version
Unrecognized OpenGL version
Unrecognized OpenGL version
Unrecognized OpenGL version
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
QQuickWidget: Failed to make context current
QQuickWidget::resizeEvent() no OpenGL context
QQuickWidget: Failed to make context current
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
Unrecognized OpenGL version
Unrecognized OpenGL version
Unrecognized OpenGL version
Unrecognized OpenGL version
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
QQuickWidget: Attempted to render scene with no context
QQuickWidget: Attempted to render scene with no context
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
QQuickWidget: Attempted to render scene with no context
QQuickWidget::invalidateRenderControl could not make context current
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 8 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 8 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (1 1 1 8)
```
3. Output with `desktop.renderingEngine=software` set (RStudio launches successfully):
```
libGL error: No matching fbConfigs or visuals found
libGL error: failed to load driver: swrast
Unrecognized OpenGL version
Unrecognized OpenGL version
```

**Comment by arnonerba** [comment]
So, I'm realizing the underlying problem here has already been discussed in https://github.com/rstudio/rstudio/issues/3942, https://github.com/rstudio/rstudio/issues/3382, and https://github.com/rstudio/rstudio/issues/3185.

@kevinushey, would it be possible to have RStudio automatically pick "software" as the rendering engine if it is launched over SSH with X11 forwarding and `Rendering engine` is set to `Auto-detect` in the GUI? Spyder had a very similar issue (https://github.com/spyder-ide/spyder/issues/7447) and now appears to default to software rendering when launched over X11. It seems that the core issue here is RStudio choosing `Desktop OpenGL` in a situation where it's not truly available.

I realize that RStudio Desktop does not officially support running over X11, but I have a large number of users who primarily interact with RStudio this way.

**Comment by kevinushey** [comment]
Is there a way for us to easily detect this case?

**Comment by arnonerba** [comment]
I think if both `$DISPLAY` and `$SSH_CONNECTION` are set it's possible to assume the user is logged in via SSH with X11 forwarding enabled. Is that a reasonable source of information that RStudio could use?

Bash example, for my own reference:
```bash
if [[ -n $DISPLAY && -n $SSH_CONNECTION ]]; then echo "x11"; fi
```

**Comment by gdevenyi** [comment]
This is still broken
